CSHARP-5626 Add evergreen script to generate CycloneDX SBOM

jasonhills-mongodb opened the pull request:
## Overview

This PR automates `sbom.json` generation and upload using a two-stage pipeline:

1. **GitHub Actions** (`sbom.yml`): Triggered when NuGet dependency files change on `main`. Regenerates `sbom.json` and opens an automated PR if the content has changed. This stage runs in GitHub Actions because it needs GitHub API access for license resolution.

2. **Evergreen** (`upload-sbom` task/variant): Triggered when `sbom.json` changes. Augments the SBOM via silkbomb and uploads it to the internal tracking system. This stage runs in Evergreen because it needs access to internal ECR and IAM roles.

## Changes

### New files
- **`MongoDB.Driver.sbom.slnf`** — Solution filter scoping SBOM generation to the 4 production packages (`MongoDB.Bson`, `MongoDB.Driver`, `MongoDB.Driver.Authentication.AWS`, `MongoDB.Driver.Encryption`). Replaces the per-package loop and `packages.sh`.
- **`.github/workflows/sbom.yml`** — GitHub Actions workflow that regenerates `sbom.json`, validates it with the CycloneDX CLI (SHA256-pinned binary), and opens an automated PR via `peter-evans/create-pull-request` when content changes.
- **`evergreen/upload-sbom.sh`** — Pulls silkbomb from ECR, runs `augment`, diffs against the previous augmented SBOM, and posts a failing task status to Evergreen if significant changes are detected (while allowing the build to continue).

### Modified files
- **`evergreen/generate-sbom.sh`** — Complete rewrite:
  - Upgraded from `cyclonedx-dotnet` 5.3.1 → **6.2.0**, which adds solution filter (`.slnf`) support, eliminating the need for Docker and the per-package loop + merge approach
  - Single `dotnet-CycloneDX` invocation on the `.slnf` instead of 4 separate invocations + Docker merge
  - `--spec-version 1.5` to prevent cyclonedx-dotnet 6.2.0 from defaulting to 1.7
  - `/p:TreatWarningsAsErrors=false` on restore to suppress NU1902/NU1903 vulnerability warnings from `MongoDB.ruleset`
  - Stable `serialNumber` (hardcoded UUID, same across all regenerations)
  - Monotonic `.version` field — only increments when SBOM content changes (compared without `metadata.timestamp` and `.version`)
  - `libmongocrypt` injected as a component after generation: version extracted from `LibMongoCryptVersion` in `MongoDB.Driver.Encryption.csproj` using `dotnet msbuild -getProperty:LibMongoCryptVersion`. Added to `MongoDB.Driver.Encryption`'s `dependsOn` in the `dependencies` graph.
  - Drops Docker dependency entirely for SBOM generation
- **`evergreen/evergreen.yml`** — Adds `upload-sbom` function, task (`tags: ["ssdlc"]`), and `sbom` build variant with `paths: [sbom.json]` trigger
- **`evergreen/download-augmented-sbom.sh`** — Fixed path bug: `cp /pwd/${SSDLC_PATH}/...` used `/pwd/` outside a Docker context; corrected to `cp "${SSDLC_PATH}/..."`

### Deleted files
- **`purls.txt`** — No longer referenced or needed.

## Addressing previous review feedback

- **PURL patching bug** (`startswith("MongoDB.")` + `join(",")` stamping wrong PURLs): Entire approach replaced. The `.slnf` causes the 4 MongoDB packages to be treated as in-solution projects, so they never appear in `components[]` and no PURL patching is needed.
- **`/pwd/` path in `download-augmented-sbom.sh`**: Fixed.
- **Environment variable comment block**: Added at top of `generate-sbom.sh`.
- **Echo messages for GitHub license resolution**: Added (`enabled for user: X` / `disabled`).
- **Docker image version/digest concern**: Docker is no longer used for SBOM generation. The `upload-sbom.sh` uses the `silkbomb:2.0` tag (consistent with other Evergreen-based silkbomb usage in the driver ecosystem).
- **`--exclude-dev` not covering `PrivateAssets="All"` packages**: Confirmed — `--exclude-dev` finds 0 dev dependencies because NuGet's `DevelopmentDependency` metadata is separate from MSBuild's `PrivateAssets`. The `--exclude-filter` list handles the 3 build-tool packages (`FxCopAnalyzers`, `NETFramework.ReferenceAssemblies`, `SourceLink.GitHub`) reliably.

## SBOM content comparison

The new output vs. the previous per-package + Docker merge approach:

| | Old (5.3.1 + docker merge) | New (6.2.0 + `.slnf`) |
|---|---|---|
| `specVersion` | 1.6 (cyclonedx-cli silently upgraded) | 1.5 ✓ |
| `serialNumber` | Random UUID each run | Stable across regenerations |
| Shape | Hierarchical: 4 MongoDB packages each with nested sub-components | Flat with `dependencies[]` graph |
| NuGet packages captured | 25 (same) | 30 (updated after merging upstream) |
| `MongoDB.Bson` / `MongoDB.Driver` as components | Yes (via ProjectReference→PURL hack) | No (correctly omitted — source projects) |
| `libmongocrypt` | Not included | Included (1.18.2), linked to `MongoDB.Driver.Encryption` |
| Docker required | Yes | No |
| Version tracking | None | Monotonic `.version`, stable `serialNumber` |

## Testing

- `dotnet-CycloneDX` run locally against the `.slnf` confirmed: 30 NuGet components, `specVersion: 1.5`, all 3 excluded packages absent, all packages have NuGet PURLs.
- `dotnet msbuild -getProperty:LibMongoCryptVersion` extracts `1.18.2` directly — no GitHub API lookup needed.
- Old script run locally for comparison: cyclonedx-cli merge step silently upgraded `specVersion` to 1.6; confirmed equivalent NuGet package coverage in both approaches.

jasonhills-mongodb committed:
CSHARP-5626: Add evergreen script to generate CycloneDX SBOM

Added a bash script to generate a 'build' lifecycle CycloneDX SBOM using
the cyclonedx-dotnet tool. The script installed a fixed version of
cyclonedx-dotnet, runs a versioned dotnet restore, and generates an SBOM
for each of the 4 MongoDB.Driver Nuget packages.

To ensure accurate results, two queries are run against each of the
.csproj files to ensure all development packages are excluded and that
any local <ProjectReference> items are marked in the SBOM as Nuget
packages. Once all 4 SBOMs are created, they are merged into a single
heirarchical SBOM file.

The file is saved as sbom.cdx.json (as opposed to the current sbom.json)
which is the preferred file extention for CycloneDX files. There is not
yet any code to commit the new SBOM to the repo. This is to allow for
evaluation of the new SBOM first without intefering with the current
workflow to upload the current static sbom.json file to Kondukto.

There is also a line added to download-augmented-sbom.sh to copy the
augmented SBOM to ./vex.cdx.json (also not yet committed to repo) after
it has been uploaded to the release artifacts bucket, as this is how we
should be storing the augmented SBOM for public consumption.

Once the SBOM generation process has been approved, I will add commit
code for both of the cdx.json files, remove th sbom.json file and update
download-augmented-sbom.sh to use the new sbom.cdx.json file.

jasonhills-mongodb committed:
Revert changes made to evergreen.yml

Undoing unintended changes to evergreen.yml

jasonhills-mongodb committed:
Merge branch 'mongodb:main' into CSHARP-5626/sbom_generation_using_cyclonedx-dotnet

adelinowona commented on evergreen/generate-sbom.sh:
```diff
@@ -0,0 +1,87 @@
+  # Patch JSON file with PURLs, as needed
+  for patch in $PURL_PATCHES; do
+    echo "Patching ${patch} with Nuget PURL"
+    contents=$(jq --arg package "$patch" --arg version "$PACKAGE_VERSION" '.components |= map(if [.name | startswith("MongoDB.")] and has("purl") == false then .purl = "pkg:nuget/\($package)@\($version)" else . end)' ${SBOM_FILE})
```
Question: If a package has multiple ProjectReferences (e.g., MongoDB.Driver references both MongoDB.Bson and MongoDB.Driver.Core), won't this logic:
  - Loop 1: Match ALL components starting with "MongoDB." that lack PURLs, assign all of them pkg:nuget/MongoDB.Bson@VERSION
  - Loop 2: All MongoDB components now have PURLs, so nothing gets patched with MongoDB.Driver.Core

  Should the condition be .name == $package instead of .name | startswith("MongoDB.")?

jasonhills-mongodb replied:
Logic has been revamped

adelinowona commented on evergreen/generate-sbom.sh:
```diff
@@ -0,0 +1,87 @@
+#!/usr/bin/env bash
+set -o errexit # Exit the script with error if any of the commands fail
+
```
```suggestion
# Environment Variables:
  #   PACKAGE_VERSION (optional) - Version to use for SBOM generation
  #                                Defaults to output of get-version.sh
  #   GITHUB_USER (optional)     - GitHub username for license resolution
  #   GITHUB_APIKEY (optional)   - GitHub API token for license resolution
```
Add a comment block to describe variables used in the script.

adelinowona commented on evergreen/generate-sbom.sh:
```diff
@@ -0,0 +1,87 @@
+  ## Run cyclonedx-dotnet
+  # Attempt GitHub license resolution only if GITHUB_USER and GITHUB_APIKEY are both set
+  if [[ -v GITHUB_USER && -v GITHUB_APIKEY ]]; then
+    github_options=(--enable-github-licenses --github-username ${GITHUB_USER} --github-token ${GITHUB_APIKEY})
```
Maybe add some echo messages to indicate whether github license resolution is enabled or not

jasonhills-mongodb replied:
Good idea. Done.

adelinowona commented on evergreen/generate-sbom.sh:
```diff
@@ -0,0 +1,87 @@
+  # There are nuances to how cyclonedx-dotnet handles <PackageReference> items in Directory.Build.props that lead to private packages being included in SBOM
+  # results even when PrivateAssets is set to "All". As a safeguard, this command lists the PackageReferences and adds the references with PrivateAssets="All"
+  # to an exclusion filter variable to be fed into cyclonedx-dotnet
+  EXCLUDE_FILTER=$(dotnet msbuild ./src/${package}/${package}.csproj -getItem:PackageReference | jq -r '[.Items.PackageReference[] | select(.PrivateAssets != null) | select(.PrivateAssets | test ("All"; "i")) | .Identity + "@" + .Version] | join(",")')
```
So the option `--exclude-dev` doesn't help with this?

jasonhills-mongodb replied:
The `--exclude-dev` only catches components with certain properties defined. The revamped approach improves on this.

jasonhills-mongodb committed:
CSHARP-5626: Rewrite SBOM generation using cyclonedx-dotnet 6.2.0 and solution filter

Replaces the per-package loop + cyclonedx-cli Docker merge approach with a single
cyclonedx-dotnet 6.2.0 invocation on a new MongoDB.Driver.sbom.slnf solution filter.

Key changes:
- MongoDB.Driver.sbom.slnf: solution filter scoping generation to the 4 production
  packages, replacing the packages.sh loop and ProjectReference PURL-patching workarounds
- generate-sbom.sh: upgraded to cyclonedx-dotnet 6.2.0; drops Docker dependency;
  adds stable serialNumber, monotonic version tracking, and libmongocrypt injection
  (version resolved from LibMongoCryptCommit via GitHub tags API, with SHA fallback)
- .github/workflows/sbom.yml: new GitHub Actions workflow that regenerates sbom.json
  and opens an automated PR when NuGet dependency files change on main
- evergreen/evergreen.yml: new upload-sbom function and task, new sbom build variant
  triggered by changes to sbom.json
- evergreen/upload-sbom.sh: new script that uploads via silkbomb to ECR
- evergreen/download-augmented-sbom.sh: fix path bug (cp used /pwd/ outside Docker context)

jasonhills-mongodb commented on evergreen/generate-sbom.sh:
```diff
@@ -0,0 +1,87 @@
+
+# Use cyclonedx-cli to merge the SBOMs into 1 hierarchical SBOM
+docker run --platform="linux/amd64" --rm -v ${PWD}:/pwd \
+  cyclonedx/cyclonedx-cli:0.28.2 \
```
You are right, though cyclonedx-cli has removed from the workflow.

jasonhills-mongodb commented on evergreen/download-augmented-sbom.sh:
```diff
@@ -21,4 +21,6 @@ echo "KONDUKTO_TOKEN=$kondukto_token" > ${PWD}/kondukto_credentials.env
\ No newline at end of file
+  augment --repo mongodb/mongo-csharp-driver --branch main --sbom-in /pwd/sbom.json --sbom-out /pwd/${SSDLC_PATH}/augmented-sbom.json
+
+cp /pwd/${SSDLC_PATH}/augmented-sbom.json ./vex.cdx.json
```
Those options are getting fed into the Python script that lives inside the container.

adelinowona replied:
Agreed the docker flags are fine — and the `cp` line outside the container is now fixed too, thanks

jasonhills-mongodb committed:
CSHARP-5626: Address Copilot review comments in generate-sbom.sh

- Use -n instead of -v for GITHUB_USER/GITHUB_APIKEY checks so that
  set-but-empty variables do not enable license resolution or auth headers
- Guard against empty MONGOCRYPT_COMMIT with an explicit error exit
- Validate GitHub API response is an array before processing to handle
  rate-limit or auth error objects gracefully
- Write libmongocrypt jq output to a temp file and mv into place rather
  than capturing into a shell variable

jasonhills-mongodb committed:
CSHARP-5626: Address Copilot review comments in upload-sbom.sh

- Add --user "$(id -u):$(id -g)" to docker run so silkbomb output files
  are owned by the invoking user rather than root
- Use mktemp for old/new/diff temp files with a trap for cleanup, avoiding
  working directory pollution and concurrent-run interference
- Add http:// scheme to localhost:2285 task status URL so curl can connect

jasonhills-mongodb commented on evergreen/generate-sbom.sh:
```diff
@@ -0,0 +1,136 @@
+MONGOCRYPT_COMMIT=$(sed -n 's/.*<LibMongoCryptCommit>\([^<]*\)<\/LibMongoCryptCommit>.*/\1/p' \
+  src/MongoDB.Driver.Encryption/MongoDB.Driver.Encryption.csproj | head -1)
+
+echo "Looking up tag for libmongocrypt commit ${MONGOCRYPT_COMMIT}"
```
Lines 67–70 add the guard that exits with a clear error message if `MONGOCRYPT_COMMIT` is empty

jasonhills-mongodb committed:
Merge remote-tracking branch 'upstream/main' into CSHARP-5626/sbom_generation_using_cyclonedx-dotnet

jasonhills-mongodb committed:
CSHARP-5626: Simplify libmongocrypt version lookup after upstream change

Upstream switched from commit-based S3 downloads (LibMongoCryptCommit +
LibMongoCryptVersion as branch prefix) to GitHub Releases downloads with
LibMongoCryptVersion as a proper semver (1.18.2). The version is now
available directly, so the GitHub tags API pagination loop is no longer
needed. Also update sbom.json to reflect the current dependency graph.

jasonhills-mongodb committed:
CSHARP-5626: Use msbuild -getProperty to extract LibMongoCryptVersion

Replaces the sed text-parsing approach with dotnet msbuild -getProperty,
which uses the MSBuild evaluation engine and correctly handles computed
or conditionally-set property values.

jasonhills-mongodb committed:
CSHARP-5626: Remove purls.txt (no longer needed)

adelinowona commented on evergreen/evergreen.yml:
```diff
@@ -222,6 +222,60 @@ functions:
       params:
         file: mo-expansion.yml
 
+  generate-sbom:
+    - command: shell.exec
+      params:
+        working_dir: "mongo-csharp-driver"
+        env:
+          GITHUB_USER: ${github_user}
+          GITHUB_APIKEY: ${github_apikey}
+          PACKAGE_VERSION: ${PACKAGE_VERSION}
+        script: |
+          ${PREPARE_SHELL}
+          ./evergreen/generate-sbom.sh
```
The generate-sbom function is defined here but I can't find anything that invokes it — no task uses func: generate-sbom, and the sbom variant only runs upload-sbom. Since SBOM generation now lives entirely in the .github/workflows/sbom.yml GitHub Actions workflow, is this Evergreen function leftover from the earlier design? If so, let's delete it; if it's intended for manual/future use, a comment saying so would help.

jasonhills-mongodb replied:
Removed the dead `generate-sbom` function (SBOM generation lives entirely in `.github/workflows/sbom.yml`)

adelinowona commented on evergreen/upload-sbom.sh:
```diff
@@ -0,0 +1,41 @@
+diff_txt=$(mktemp)
+trap 'rm -f "$old_json" "$new_json" "$diff_txt"' EXIT
+
+if [ -f ./augmented.sbom.json ]; then
+  jq -S 'del(.metadata.timestamp)' ./augmented.sbom.json > "$old_json"
+else
+  echo '{}' > "$old_json"
+fi
+jq -S 'del(.metadata.timestamp)' ./augmented.sbom.json.new > "$new_json"
```
The "significant changes" check diffs augmented.sbom.json.new against augmented.sbom.json, but nothing in the upload-sbom task ever creates or downloads augmented.sbom.json (the script only writes the .new file, and the task runs func: upload-sbom alone). So old_json always falls into the echo '{}' branch and the diff reports significant changes on every run, making the guard effectively constant-true. Was the intent to download the previously-augmented SBOM as the baseline first (analogous to download-augmented-sbom.sh), or to persist .new back as the baseline after upload?

jasonhills-mongodb replied:
The `augment` command no longer used/needed. Instead using `upload` without a diff.

adelinowona commented on evergreen/upload-sbom.sh:
```diff
@@ -0,0 +1,41 @@
+silkbomb="901841024863.dkr.ecr.us-east-1.amazonaws.com/release-infrastructure/silkbomb:2.0"
+docker pull "${silkbomb}"
+
+silkbomb_augment_flags=(
+  --repo mongodb/mongo-csharp-driver
+  --branch "${branch_name}"
+  --sbom-in /pwd/sbom.json
+  --sbom-out /pwd/augmented.sbom.json.new
+  --no-update-sbom-version
+)
+
+docker run --rm -v "$(pwd):/pwd" \
+  --user "$(id -u):$(id -g)" \
+  --env 'AWS_ACCESS_KEY_ID' --env 'AWS_SECRET_ACCESS_KEY' --env 'AWS_SESSION_TOKEN' \
+  "${silkbomb}" augment "${silkbomb_augment_flags[@]}"
```
This runs the same silkbomb `augment` command as our existing `download-augmented-sbom.sh`, but the `upload-sbom` function assumes the silkbomb role and forwards only AWS creds + branch_name + DOCKER_CONFIG — it never runs `fetch-kondukto-token.sh` and never passes` --env-file kondukto_credentials.env / KONDUKTO_TOKEN`.

The download path goes through` kondukto_role_arn → fetch-kondukto-token.sh → --env-file` for the identical `augment` call, which suggests `augment` needs the `Kondukto` token to reach `Kondukto`. Does the silkbomb IAM role somehow cover `Kondukto` auth (e.g. silkbomb self-fetches the token from Secrets Manager), or is the f`etch-kondukto-token.sh + --env-file` step missing here?

jasonhills-mongodb replied:
Good callout. When running in Evergreen, `silkbomb` only needs AWS credentials. It retrieves the keys for Kondukto and Dependency Track automatically.
